### PR TITLE
Feature/check macro is referenced

### PR DESCRIPTION
**Pull request by benalloum**

### CODE REVIEW

File: HOOKS.md [hunk 1]
```diff
@@ -889,6 +892,27 @@ You want to make sure that all specified columns in the properties files (usuall
 
 -----
 
+### `check-source-folder-constraint`
```
Comment (benoitgoujon) [inline HOOKS.md:895]: The name of this hook does not match the conventions used in other hooks.

I would prefer a name like `check-sources-are-not-in-directory` or sth like that but if you look at other names, there's always a verb. Consistency is key.

File: pre_commit_dbt/check_macro_is_referenced.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+if __name__ == "__main__":
+     exit(main())
+
+# # Opening JSON file
+# home_path=str(Path.home())
+# manifest_path=f'{home_path}/Downloads/jaffle_shop-main/target/manifest.json'
+# f = open(manifest_path)
+# data = json.load(f)
+# path_macro = [f'{home_path}/Downloads/jaffle_shop-main/macros/cents_to_dollars.sql']
+# has_reference(path_macro, data)
```
Comment (benoitgoujon) [inline pre_commit_dbt/check_macro_is_referenced.py:68]: No dead code in the main branch!

File: pre_commit_dbt/check_macro_is_referenced.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            macros_transformed_names = map(lambda x:x.split(".")[-1], macros_as_declared_in_manifest) # maps original list elements which have the following format macro.[project_name].[macro_name] to macro_name
+            if macro_name in macros_transformed_names:
+                found = True
+                break
+            else:
+                continue
+        if not found:
+            print(f'The macro {macro_name} was not referenced in any model.')
+            status_code = 1
+    return status_code
```
Comment (benoitgoujon) [inline pre_commit_dbt/check_macro_is_referenced.py:43]: The complexity of this block of code is high. It is quite hard to read.

Can you explain to me in plain english what it does? I would like to try to simplify if possible. Ideas from other team members are of course welcome.

As a starting point, I am not sure we need the `else: continue` block. It adds cognitive loads. We don't care about 1 ms performance for this project, we care about readability.

File: pre_commit_dbt/check_source_folder_constraint.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import argparse
+from pathlib import Path
+from typing import Optional
+from typing import Sequence
+
+from pre_commit_dbt.utils import add_filenames_args
+from pre_commit_dbt.utils import get_source_path
+
+"""
```
Comment (benoitgoujon) [inline pre_commit_dbt/check_source_folder_constraint.py:9]: DOn't think this is useful. Docstrings are useful for functions that are reused multiple times like helper functions.

File: pre_commit_dbt/check_macro_is_referenced.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+from pre_commit_dbt.utils import add_manifest_args
+from pre_commit_dbt.utils import get_filenames
+from pre_commit_dbt.utils import get_json
+from pre_commit_dbt.utils import get_macro_schemas
+from pre_commit_dbt.utils import get_macro_sqls
+from pre_commit_dbt.utils import get_macros
+from pre_commit_dbt.utils import JsonOpenError
+
+
+def has_reference(paths: Sequence[str], manifest: Dict[str, Any]) -> int:
```
Comment (benoitgoujon) [inline pre_commit_dbt/check_macro_is_referenced.py:20]: This function would deserve a dedicated unit test given its complexity.

**Comment by larryandria** [comment]
It seems that you've also implemented the 'source folder constraint' hook but there is already another PR for that.

**Comment by benoitgoujon** [comment]
> It seems that you've also implemented the 'source folder constraint' hook but there is already another PR for that.

Absolutely! Thanks Larry. I was wondering why we had 2 hooks in this PR.
